var/spack/repos/builtin/packages/python/package.py

```diff
@@ -949,10 +949,37 @@ def setup_dependent_build_environment(self, env, dependent_spec):
         if not is_system_path(path):
             env.prepend_path('PATH', path)
 
-        for d in dependent_spec.traverse(deptype=('build', 'run', 'test'), root=True):
-            if d.package.extends(self.spec):
-                env.prepend_path('PYTHONPATH', join_path(
-                    d.prefix, self.site_packages_dir))
+        # Add installation prefix to PYTHONPATH, needed to run import tests
+        prefixes = set()
+        if dependent_spec.package.extends(self.spec):
+            prefixes.add(dependent_spec.prefix)
+
+        # Add direct build/run/test dependencies to PYTHONPATH,
+        # needed to build the package and to run import tests
+        for direct_dep in dependent_spec.dependencies(deptype=('build', 'run', 'test')):
+            if direct_dep.package.extends(self.spec):
+                prefixes.add(direct_dep.prefix)
+
+                # Add recursive run dependencies of all direct dependencies,
+                # needed by direct dependencies at run-time
+                for indirect_dep in direct_dep.traverse(deptype='run'):
+                    if indirect_dep.package.extends(self.spec):
+                        prefixes.add(indirect_dep.prefix)
```

Comment:
Don't recurse here, but just set PYTHONPATH for this package only.

Reply:
This won't work without recursion. If A imports B and B imports C and C imports D, all packages need to be present in the PYTHONPATH.

Reply:
See https://github.com/spack/spack/blob/c85c725b082756d2b18ddfb1f8995a84f1bf40cf/lib/spack/spack/build_environment.py#L787-L788, it's taken care of.

Reply:
@haampie I tried removing this but it causes every build to fail. If I look at `spack-build-env.txt`, only the current package gets added to `PYTHONPATH`, not any of its dependencies. This means that `python -m pip` won't even work because it can't find pip. Could it be because we're setting this in `setup_dependent_build_environment` instead of `setup_run_environment`? This reminds me of the discussion in https://github.com/spack/spack/discussions/8639.

Reply:
PyPip is not a PythonPackage

Reply:
Yes, but it extends Python, so these settings should still be applied. Not only is pip missing, all dependencies are missing in PYTHONPATH.

Reply:
Right, I just noticed you put this logic in Python's package.py, not in the build system. Still I think it's better to make the python package build system define a generic setup_dependent_build_environment which sets the variable per package. And later we can replace that with setup_run_environment indeed.

Reply:
along the lines of

```diff
diff --git a/lib/spack/spack/build_systems/python.py b/lib/spack/spack/build_systems/python.py
index 65bfd09eaf..2400a07f7c 100644
--- a/lib/spack/spack/build_systems/python.py
+++ b/lib/spack/spack/build_systems/python.py
@@ -48,6 +48,10 @@ class PythonPackage(PackageBase):
 
     py_namespace = None
 
+    def setup_dependent_build_environment(self, env, dependent_spec):
+        python_dir = 'python{0}'.format(self.spec['python'].version.up_to(2))
+        env.prepend_path('PYTHONPATH', os.path.join(self.prefix, 'lib', python_dir, 'site-packages'))
+
     @staticmethod
     def _std_args(cls):
         return [
diff --git a/var/spack/repos/builtin/packages/py-pip/package.py b/var/spack/repos/builtin/packages/py-pip/package.py
index e4cb9b2bb8..3d8e89164b 100644
--- a/var/spack/repos/builtin/packages/py-pip/package.py
+++ b/var/spack/repos/builtin/packages/py-pip/package.py
@@ -49,6 +49,10 @@ def install(self, spec, prefix):
         args = [os.path.join(whl, 'pip')] + std_pip_args + ['--prefix=' + prefix, whl]
         python(*args)
 
+    def setup_dependent_build_environment(self, env, spec):
+        python_dir = 'python{0}'.format(self.spec['python'].version.up_to(2))
+        env.prepend_path("PYTHONPATH", os.path.join(self.prefix, 'lib', python_dir, 'site-packages'))
+
     def setup_dependent_package(self, module, dependent_spec):
         pip = self.spec['python'].command
         pip.add_default_arg('-m')
diff --git a/var/spack/repos/builtin/packages/py-wheel/package.py b/var/spack/repos/builtin/packages/py-wheel/package.py
index 12ca492976..3d9e76f5a4 100644
--- a/var/spack/repos/builtin/packages/py-wheel/package.py
+++ b/var/spack/repos/builtin/packages/py-wheel/package.py
@@ -4,7 +4,7 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
 from spack import *
-
+import os
 
 class PyWheel(Package):
     """A built-package format for Python."""
@@ -30,6 +30,10 @@ class PyWheel(Package):
     depends_on('python@2.6:2.8,3.2:', type=('build', 'run'))
     depends_on('py-pip', type='build')
 
+    def setup_dependent_build_environment(self, env, spec):
+        python_dir = 'python{0}'.format(self.spec['python'].version.up_to(2))
+        env.prepend_path("PYTHONPATH", os.path.join(self.prefix, 'lib', python_dir, 'site-packages'))
+
     def install(self, spec, prefix):
         # To build wheel from source, you need setuptools and wheel already installed.
         # We get around this by using a pre-built wheel, see:
diff --git a/var/spack/repos/builtin/packages/python/package.py b/var/spack/repos/builtin/packages/python/package.py
index 80b0723820..00cb9d96c1 100644
--- a/var/spack/repos/builtin/packages/python/package.py
+++ b/var/spack/repos/builtin/packages/python/package.py
@@ -949,38 +949,6 @@ def setup_dependent_build_environment(self, env, dependent_spec):
         if not is_system_path(path):
             env.prepend_path('PATH', path)
 
-        # Add installation prefix to PYTHONPATH, needed to run import tests
-        prefixes = set()
-        if dependent_spec.package.extends(self.spec):
-            prefixes.add(dependent_spec.prefix)
-
-        # Add direct build/run/test dependencies to PYTHONPATH,
-        # needed to build the package and to run import tests
-        for direct_dep in dependent_spec.dependencies(deptype=('build', 'run', 'test')):
-            if direct_dep.package.extends(self.spec):
-                prefixes.add(direct_dep.prefix)
-
-                # Add recursive run dependencies of all direct dependencies,
-                # needed by direct dependencies at run-time
-                for indirect_dep in direct_dep.traverse(deptype='run'):
-                    if indirect_dep.package.extends(self.spec):
-                        prefixes.add(indirect_dep.prefix)
-
-        # Packages may be installed in platform-specific or platform-independent
-        # site-packages directories
-        site_packages_dirs = set()
-        for plat_specific in ['true', 'false']:
-            try:
-                site_packages_dirs.add(
-                    self.config_vars['python_lib'][plat_specific]['false']
-                )
-            except KeyError:
-                site_packages_dirs.add(self.default_site_packages_dir)
-
-        for prefix in prefixes:
-            for site_packages_dir in site_packages_dirs:
-                env.prepend_path('PYTHONPATH', os.path.join(prefix, site_packages_dir))
-
         # We need to make sure that the extensions are compiled and linked with
         # the Spack wrapper. Paths to the executables that are used for these
         # operations are normally taken from the sysconfigdata file, which we
```

Reply:
This would also need to be done for the hundreds of other packages that extend Python but don't subclass from `PythonPackage`. I don't see a clear benefit of this over the current solution.

Reply:
the upside is that we don't duplicate the traversing logic

Reply:
Yes, if we remove these 10 lines of code we can duplicate the other 5 lines of code hundreds of times. I think the additional code duplication outweighs any benefit of preventing packages from setting env vars for their deps. Until Spack properly handles recursive env var setup I think it's best to keep things as is and discuss this in a follow-up PR.